### ISSUE TITLE
Upgrade to etcd v3 plan

### ISSUE THREAD

**Issue by hongchaodeng** [kubernetes/kubernetes#22448]
The etcd team has do a lot work to make etcd v3 works efficiently in large scale. Recently we have done a prototype that successfully ran k8s with etcd v3 :) Now I think it’s time for us to start initial discussion and work on the upgrade of etcd in Kubernetes.
## What’s new in etcd v3

First, I want to point out the major benefits of etcd v3:
- We would get rid of relist problem in most cases. etcd3 is able to keep millions of keys. So we can easily keep keys of last 1 hour. Now etcd3 can send progress notifications to watchers. If a client disconnects for an hour, we probably should let it die and doing a full recovery instead of a relist.
- One lease per component. We would use only one lease client to keep alive all TTL keys. The lease has its own ID and is not tied to any client session so that we can migrate it easily. The cost of lease is also much lower than previous TTL keys.
- Efficient watch: Use gRPC insead of HTTP + JSON. Multiple watchers per one TCP connection and very efficient encoding.
- Mini-transaction (Txn): simplify multiple key creation logic.
- Reliability: it has much lower memory footprint, nonetheless it's able to run smoothly with GBs of keys in mem.
- Other improvements like consistent range (pagination) over large number of keys, being able to watch all deletions when delete a prefix, etc.
## Attack Plan

@wojtek-t pointed out that “pkg/storage/etcd” package encapsulates all etcd related code. We appreciate the community work on this! I’m thinking about two ways to accomplish v3 upgrade:

The first way is that we can reuse the interface and rewrite the implementation in etcd3. This is the easy way. We can get the work done easily and quickly. The disadvantage is that the interface resembles that of etcd2. We might need to do some workaround for incompatible changes (ttl to lease). In the long term, we also need to add methods to take advantage of new features like Txn. 

The other way is that we can design a new interface for etcd3. We can keep v3 API in mind and make it efficiently use etcd in first place. However, this requires more work and slows things down initially.

Either way we can have it done in parallel with existing storage layer and add a flag to switch to etcd3.

TBH, I’m not leaning to either way. I saw a lot of interest from the community and it’s important for us to coordinate and move as fast as we can. Any thoughts are welcome!

**Comment by hongchaodeng** [comment]
@wojtek-t @gmarek @timothysc @lavalamp @smarterclayton @xiang90 @HardySimpson 

This is a fork from #20504. Besides introducing the benefits of etcd3, I want to discuss specifically what we are going to deal with storage.Interface and etcdHelper struct.

Please feel free to "at" more people if needed.

**Comment by ncdc** [comment]
cc @kubernetes/rh-cluster-infra

**Comment by krousey** [comment]
cc @kubernetes/sig-api-machinery

**Comment by timothysc** [comment]
@hongchaodeng You can write a totally parallel implementation behind the storage interface and leave the existing one as is.  Startup could be conditionalized behind a factory.  I think we'd have to migrate the capabilities through the storage interface eventually, but this ripples all through the registry.  

Lastly, someone is going to need to update sky-dns, where last I looked, we were using a really old version.

**Comment by alfred-huangjian** [comment]
cc @kubernetes/huawei

**Comment by xiang90** [comment]
@timothysc 

> re: gRPC, that's an orthogonal issue.

etcd uses gRPC for client/server communication. So in the context of etcd + k8s, this is relevant. This is not API sever + other components through gRPC,

**Comment by wojtek-t** [comment]
I agree with @timothysc - we should start with implementing v3 etcd implementation behind currently existing interface. Once we have this working, we can start extending/changing it to make benefits of etcd v3, but that's a second step in my opinion.

Regarding gRPC - I think there are few points here:
- does etcd v3 support only gRPC or also JSON? I though both, am I right?
- eventually, we definitely want to migrate to gRPC the communication etcd <-> apiserver; but I'm not sure if starting with gRPC will not cause more troubles in migration path; we need to be careful about it
- how much additional work would be start with JSON implementation first, and then rewrite it to gRPC?

**Comment by smarterclayton** [comment]
We probably want gRPC early so that we can send binaries with protobuf,
rather than the escaped form in JSON.  I'd rather move away from assuming
JSON now in the generic storage.  Is the only concern about gRPC that it's
going to complicate our internal code path?

On Fri, Mar 4, 2016 at 2:51 AM, Wojciech Tyczynski <notifications@github.com

**Comment by wojtek-t** [comment]
I think that in the generic storage we don't have any assumption about JSON - it is only assumed in the etcd implementation of storage.Interface and only at the implementation level.

I'm definitely not concerned about code complication - I'm more worried if it won't complicate upgrade path (i.e. from 1.2 to 1.3 supporting etcd v3). 
But I'm fine with starting with gRPC too if you think that won't be a problem.

**Comment by xiang90** [comment]
@wojtek-t @smarterclayton 

gRPC wont be a problem. etcd client does all the work. kubernetes storage layer should not even know about it unless you do not want to introduce any gRPC dependencies.

**Comment by wojtek-t** [comment]
I agree that storage interface shouldn't expose any gRPC-related things

**Comment by lavalamp** [comment]
> You can write a totally parallel implementation behind the storage interface and leave the existing one as is. Startup could be conditionalized behind a factory.

Yes. It is an absolute requirement that both etcd 2 and etcd 3 work at the same time. We have etcd options plumbed through to the creation already.

gRPC is a total red herring here. Of course we should talk gRPC to etcd 3. That has zero effect on anything build on top of the storage interface.

**Comment by davidopp** [comment]
> Mini-transaction (Txn): simplify multiple key creation logic.

Can you say more about the semantics of "mini-transaction"? This sounds like it might be different from what we were assuming etcd v3 would provide, but I'm not sure.

**Comment by hongchaodeng** [comment]
> Can you say more about the semantics of "mini-transaction"? 

Please see [txn.go](https://github.com/coreos/etcd/blob/16c35167dfd026c57e993876bcd7b5e0a1d08df9/clientv3/txn.go#L33).

It's a single if-else layer and supports get, put ops in a single operation.

What's the semantics you were expecting? Is there any concern on it?

**Comment by davidopp** [comment]
No, that looks good. Thanks!

**Comment by hongchaodeng** [comment]
>  we should start with implementing v3 etcd implementation behind currently existing interface.

I send a PR #22604 as a follow up on the discussion. The change is simple but I hope it will enable more folks to start writing the code.

I have a rough plan of what we are gonna to do next:

1. Implement all KV methods (Get, Set, GuaranteedUpdate, Create, Delete, etc.) without ttl.
2. ttl is quite tricky in v3. We might need to support per-key ttl for BC while v3 encourages per client ttl. Once 1) is done, we can implement per key lease management.
3. implement watch.

Once we have initial implementation, we can further discussion testing and improvement plans. What do you guys think?

**Comment by AdoHe** [comment]
A really big TODO, I would help with KV methods.

**Comment by wojtek-t** [comment]
@hongchaodeng - yes that plan SGTM, as long as this is another implementation not replacing the currently existing one (actually this is a requirement anyway).

**Comment by hongchaodeng** [comment]
Folks.
I'm currently having trouble in https://github.com/kubernetes/kubernetes/pull/22604.

The test failure is reproducible by running "MetricsGrabber" and it seemed to grab etcd server metrics as well. From the log:
```
MetricsGrabber
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/metrics_grabber_test.go:158
  should grab all metrics from API server. [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/metrics_grabber_test.go:100

  Expected
      <sets.String | len:39>: {
          "etcd_storage_events_total": {},
          "etcd_storage_db_compaction_pause_duration_milliseconds_sum": {},
          "etcd_storage_pending_events_total": {},
          "etcd_snapshot_save_marshalling_durations_seconds_sum": {},
          "etcd_storage_index_compaction_pause_duration_milliseconds_bucket": {},
          ....
      }
  to be empty
```

Anyone have idea in this or can help do anything to resolve it? This is currently blocking me from moving forward.

**Comment by hongchaodeng** [comment]
I have found the problem.

Digging out apiserver dependencies with the new change reveals that it imports the following:
```
github.com/coreos/etcd/auth
github.com/coreos/etcd/client
github.com/coreos/etcd/clientv3
github.com/coreos/etcd/compactor
github.com/coreos/etcd/discovery
github.com/coreos/etcd/error
github.com/coreos/etcd/etcdserver
github.com/coreos/etcd/etcdserver/api/v3rpc
github.com/coreos/etcd/etcdserver/etcdhttp/httptypes
github.com/coreos/etcd/etcdserver/etcdserverpb
github.com/coreos/etcd/etcdserver/stats
github.com/coreos/etcd/lease
github.com/coreos/etcd/lease/leasehttp
github.com/coreos/etcd/lease/leasepb
...
github.com/coreos/etcd/raft
github.com/coreos/etcd/raft/raftpb
github.com/coreos/etcd/rafthttp
github.com/coreos/etcd/snap
github.com/coreos/etcd/snap/snappb
github.com/coreos/etcd/storage
github.com/coreos/etcd/storage/backend
github.com/coreos/etcd/storage/storagepb
github.com/coreos/etcd/store
github.com/coreos/etcd/version
github.com/coreos/etcd/wal
github.com/coreos/etcd/wal/walpb
```

Metrics are registered in init() func in those packages.

/cc @xiang90

**Comment by hongchaodeng** [comment]
I will try to come up with a solution tomorrow. If anyone in other timezone would like to dig into it, that would be great :)

**Comment by AdoHe** [comment]
@hongchaodeng maybe you can add those metrics to `KnownApiServerMetrics`, since etcd register metrics at `init` function.

**Comment by hongchaodeng** [comment]
ref: https://github.com/coreos/etcd/issues/4771

@AdoHe Those aren't api server metrics. I even suspect "rest_client_request_xxx" should not be either.

**Comment by AdoHe** [comment]
@hongchaodeng this is just some quick fix, AFAIK the `init` functions can not be skipped, seems you have fixed this on etcd side, great job :+1:

**Comment by hongchaodeng** [comment]
Hi folks,

I wanna give some updates and also coordinates.

I consider the KV methods completed in #23387. While waiting for review cycles, I want to do two things in parallel:
- ttl manager. As discussed above, we need to support per-key ttl for BC. I'm implementing a ttl manager roughly like this:
```Go
type TTLManager interface {
  SetTTL(key, ttl) error
  GetTTL(key) (ttl, expiration, error)
}
```
- watcher. Both Watch() and WatchList() returns a [watcher interface](https://github.com/kubernetes/kubernetes/blob/8e33cbfa281b7e464100a052c9408a1947ec29d5/pkg/watch/watch.go#L26-L35). I'm expecting a watcher to be implemented and returned. The interface goes like:
```Go
func NewWatcher(etcdClient, filter, other stuff needed...) Watcher {...}

type Watcher interface {
  // Run starts watching the keys in background and returns watched event in result chan
  Run(ctx, key, index) error
  // watch.Interface methods
  Stop()
  ResultChan() <-chan Event
}
```

I am implementing the ttl manager. If anyone else is interested in writing the code for Watcher, please take it over and do it now. I will review, help do anything if needed. I'm expecting both to be finished this week before April. Thanks in advance!

**Comment by hongchaodeng** [comment]
Submitted a PR of watch in https://github.com/kubernetes/kubernetes/pull/23694

We might deal with TTL later. It's only used in event storage and not critical.

**Comment by hongchaodeng** [comment]
Submitted a PR of TTL in https://github.com/kubernetes/kubernetes/pull/23995

We can do some parallel review and collect feedback early.

**Comment by hongchaodeng** [comment]
Have one final PR to deal with compaction in #25010.

Update:
I'm doing some work in parallel to ensure e2e tests passed.

**Comment by AdoHe** [comment]
Good job 👍 @hongchaodeng anything I can help?

**Comment by hongchaodeng** [comment]
@AdoHe 
We are a little behind timeline and have to move fast. I can think of a few things that would be helpful to us:
- Reviews on PRs. Anything that could make the code better would be welcome. If it's not blocking you can also write the code and raise it up in another PR.
- We are close to finishing initial code. But we need to run a lot of test (mostly e2e test) to ensure correctness and performance. Updates should be sync-ed here, of course. Either help run the test or fix issues found will be really helpful.
- Enable etcd3 in cluster scripts (i.e. gce, aws, coreos, etc.)
- Setting up jenkins jobs. Currently all jenkins jobs are run against etcd2. We need to run jenkisn jobs for etcd3 to prevent regression or code changes that break testing.

**Comment by wojtek-t** [comment]
Any help with e2e test would be super helpful. Currently even if we have code, we are not merging almost anything, because of flakes.

**Comment by hongchaodeng** [comment]
Re. cluster setup script, I open an issue: #25286

**Comment by hongchaodeng** [comment]
Updates on current status:
- I consider it code complete. Users should be able to run k8s with etcd3 backed storage.
- We have set up [jenkins jobs](https://github.com/kubernetes/test-infra/blob/master/jenkins/job-configs/kubernetes-jenkins/kubernetes-e2e.yaml) to run e2e testing nightly. If there are any commit that breaks the code, we should be able to know. Might be great if we can have this in upstream jenkins.
- Migration. We have worked out tools to do general (not limited to k8s) migration. Docs are still under review. We will talk about how to migrate in k8s and discuss it in upstream issue next Monday.

**Comment by timothysc** [comment]
> We will talk about how to migrate in k8s and discuss it in upstream issue next Monday. 

@hongchaodeng could I get an invite to said discussion.

**Comment by m1093782566** [comment]
Awesome work! I have built API Server 1.3 + etcd 3.0 in my envrionment and runed e2e test.

**Comment by xiang90** [comment]
@m1093782566 Does 1.3 + etcd3 work for you?

**Comment by m1093782566** [comment]
Yeah, I have passed e2e density test in a 50-node cluster. I am about to run it in a bigger cluster.

**Comment by wojtek-t** [comment]
@m1093782566 - did you also run some correctness tests to? density is mostly a performance tests and it may potentially not catch some correctness issues.

**Comment by m1093782566** [comment]
Not yet, but I will try. I plan to test API Server 1.3 + etcd 3.0 in a 3000-node cluster to see their enhancement in performance.

**Comment by wojtek-t** [comment]
@m1093782566 - note that we don't support 3000-nodes at the level of master components. So those experiments won't be very informative.

**Comment by kevin-wangzefeng** [comment]
@m1093782566 better to have some correctness tests like conformance on a cluster at reasonable size i.e. 1000/2000 nodes, before you moving to a larger scale.

**Comment by xiang90** [comment]
@kevin-wangzefeng Can you help to run a conformance test on a large cluster for etcd3 + k8s?

**Comment by kevin-wangzefeng** [comment]
@xiang90 - I have not got time to do this work myself yet, actually @m1093782566 is working with us, I will discuss with him off line :)

**Comment by timothysc** [comment]
We've run into some issues, and are digging.

**Comment by lavalamp** [comment]
What's the next step on this?

**Comment by wojtek-t** [comment]
@lavalamp 
IIUC, one thing we are waiting for is this one:
https://github.com/kubernetes/kubernetes/issues/20504#issuecomment-231274366

However, it would be great if we can figure our testing plan before that.

Once we are ready, I'm happy to setup up a Jenkins job that will be running e2e tests with etcd3 underneath. However, upgrade is something that I worry about. And it would be good if someone could write down some more detailed testing plan for it...

**Comment by wojtek-t** [comment]
@xiang90 @hongchaodeng ^^

**Comment by lavalamp** [comment]
Maybe it's time for us to find a googler to own getting this tested and
productionized. @davidopp do you have anyone free?

**Comment by wojtek-t** [comment]
I talked with @davidopp about it and we probably want have someone to help with that, but given that in 1.4 release we probably won't increase scalability (we are targetting 5000 nodes for 1.5), this is not critical for 1.4. That said, we may not be able to find anyone with free cycles in the next few weeks.

**Comment by wojtek-t** [comment]
@xiang90 @hongchaodeng - one more large thing that I think we are completely missing now are scripts. Currently apiserver can be configured to talk to etcd3, but there is:
- no way to start the cluster with etcd3 (using kube-up)
- configure apiserver based on that (in kube-up)
- and stuff like that

If I am not missing something, would you guys will be able to do this piece?

**Comment by xiang90** [comment]
@wojtek-t @lavalamp @hongchaodeng 

Most of the etcd3 support code is there. We did some very basic testing: 

- unit test inside storage/etcd3 pkg
- customized e2e testing with etcd3 that runs every night

Redhat is helping to run etcd3 with real workload.
We have a etcd2 -> etcd3 upgrade plan. We will publish it next week or so.

Next steps:

The short term goal is to make etcd3 the default backend. We think it can be done in 1.4 for fresh cluster at least.

- Setup script: Here is the issue for scripts @wojtek-t mentioned: https://github.com/kubernetes/kubernetes/issues/25286. Let's discuss the related thing in that PR.

- Multiple API Server Support PR: https://github.com/kubernetes/kubernetes/pull/28202

- Update etcd3 dependency: https://github.com/kubernetes/kubernetes/issues/20504

Besides all these, I do not know what we should do to make etcd3 is the default/production ready backend.

I agree with @lavalamp that some googlers should be involved to test/evaluate it. Redhat people are testing/evaluating it. And @kubernetes/huawei people are also probably testing it. We (Hongchao and I) would love to fix any issues you found, improve efficiency or help with anything.

It would be great if anyone can come up with concrete criteria for making etcd3 backend stable. Hongchao and I do not have enough experience operating kuberenetes cluster.

For longer term, we are trying to make the etcd3 backed storage layer efficient and reliable. We are working on: 

- add both index for cached objects and kubernetes watchers

- access past version of keys directly from etcd

- direct watchers to etcd directly if it requests from an old version

- more efficient update the object without rewriting it as a whole

- more efficient to  encode/decode objects

- and many more...

**Comment by wojtek-t** [comment]
During designing more native support of HA in kubernetes, we came up with a completely different procedure, that can save us a lot of testing & coding time.

Basically, I would like to benefit from the fact, that it's possible to create an etcd cluster that consists from etcd2 and etcd3 instances. What is more, you can do rolling update from etcd2 to etcd3. And this is exactly what we need.

So the proposal for the upgrade would be:
- we close all kube components (apiserver, scheduler, ...)
- only etcd2 stays up
- we start a new instance of etcd3 (on the same machine) and configure it so that it joins the existing etcd2
- it should propagate the whole state from the already existing etcd2 instance
- once this is done, we can simply shut down the old etcd2 instance and we have a working etcd3
- now we can follow the upgrade as we are doing it now, and we can start etcd3 from the data directory  for which the etcd3 we already started wrote its data

Advantages:
- we don't need to write any additional code to our components - it's only a matter of starting the ones that we already have with correct flags
- rollback works exactly the same way !
- we don't need to test some additional tools that we create - we simply need to test the upgrade procedure itself, because we are not using anything dedicated

@xiang90 @hongchaodeng  @lavalamp @smarterclayton - thoughts?

**Comment by timothysc** [comment]
@wojtek-t That's not going to work... I'll let @xiang90 and @hongchaodeng explain.

**Comment by hongchaodeng** [comment]
If you just want to upgrade the cluster without using any V3 API, the rolling upgrade will still work.

Basically etcd3 has two separate data store under the hood: one for v2, and one for v3. In order to use V3 API, users have to do both client migration and data migration.
- Client migration is to use v3 API in application. We can't access v3 data without using v3 API.
- Data migration is to convert v2 data into v3 data. This can be done either offline or online. For offline migration, we have supported that in `etcdctl migrate` and is gonna publish soon a more detailed document for it.

**Comment by wojtek-t** [comment]
>
If you just want to upgrade the cluster without using any V3 API, the rolling upgrade will still work.

Yeah - but that would be enough for us in the first step. We can start using V3 API in the next step (release after), right? And that would probably simplify a lot...

**Comment by xiang90** [comment]
@wojtek-t Upgrading etcd, migrating etcd data from v2 to v3, using etcd v3 API are separate issues. Yes, you can do it step by step. And I suggest you to do it step by step.

**Comment by wojtek-t** [comment]
@xiang90 - so I think I'm missing one bit of data; so is etcd3 able to work on both:
- data in the old format
- data in the new format
?
If that's true, that would explain a lot (and then my proposal is useless, because that would mean I misundersood things).

**Comment by xiang90** [comment]
@wojtek-t etcd v2 and v3 API manipulate on totally diffrent data set. They are different storage engine under the hood. Existing v2 data will not be accessible through v3 API.

**Comment by wojtek-t** [comment]
Sorry - I wasn't clear.

I understand that etcdv3 operates on different set of data.

But etcd 3.0 binary exposes also v2 API. So my question is whether data exposed via v2 API is different that those exposed via v3 API.

So assume that I start etcd3.0 binary from the old data (coming from etcd2.2 binary). This will work right? And this data will be accessible via v2 API of etcd3.0 binary, but no via its v3 API. Am I right?

**Comment by xiang90** [comment]
> So assume that I start etcd3.0 binary from the old data (coming from etcd2.2 binary). This will work right? And this data will be accessible via v2 API of etcd3.0 binary, but no via its v3 API. Am I right?

Yes. Right.

**Comment by wojtek-t** [comment]
OK - so that's the bit I was missing. Thanks a lot for clarifications!

**Comment by eparis** [comment]
$MOSTLY_USELESS_COMMENT

Does that mean maybe a good approach will be

- Stop all etcd clients (api-server, scheduler, et al)
- Add etcd3 member to cluster
- remove all etcd2 memebers
- Tell etcd3 member to upgrade all data to etcd3 (`etcd-migrate`)
- restart etcd clients

So it's the same as @wojtek-t but includes the `upgrade` component mentioned? Is that `upgrade` part hard?

**Comment by wojtek-t** [comment]
@xiang90 - in the above scenario described by @eparis , can we (instead of adding etcd3 member to the cluster), simply:
- stop all clients [actually apiserver is only component talking to etcd]
- shut down etcd2
- start etcd3 (from the old data) [IUC, this would mean that etcd3 will not support v3 API, but all the data would be there using v2 API]
- then do the migration
- then start all the clients

**Comment by xiang90** [comment]
@wojtek-t Your approach is what I would suggest! I would even suggest you to separate the migration and upgrade phase on the schedule... They do not have to be tightly coupled.

**Comment by wojtek-t** [comment]
@xiang90 - I think the main point to this discussion is whether we would like to do online migration or offline migration. If the former, I'm fine with separating it. If the later, it probably should be done when we already have components down.

**Comment by xiang90** [comment]
@wojtek-t 

It does not really matter. Separating them can reduce risk + cost in my opinion.

Phase 1 will always be online. For phase 2, you can either choose online or offline based on your requirement and schedule.

Online phase1

- shut down etcd2
- start etcd3 (from the old data) [IUC, this would mean that etcd3 will not support v3 API, but all the data would be there using v2 API]

Offline phase 2

- stop all clients [actually apiserver is only component talking to etcd]
- do the migration
- then start all the clients

Online phase 2 

see: https://github.com/coreos/etcd/blob/master/Documentation/op-guide/v2-migration.md#online-migration

**Comment by timothysc** [comment]
I'd lean towards offline migration for safety.

**Comment by xiang90** [comment]
@timothysc Yes. I prefer that one too.

**Comment by wojtek-t** [comment]
+1 for offline

**Comment by wojtek-t** [comment]
@xiang90 - IIUC, currently it is not even possible to run kubernetes using etcd3 using kube-up script. Is this something you (or someone else) is going to work on soon?

**Comment by xiang90** [comment]
@wojtek-t I do not have enough bandwidth (and expertise) to work on the setup script. It would be great if someone else can work on it. Also, we want to run e2e test with v3 api (starting api server with flag=etcd3). We have been doing this for the last two months or so. It would be great if upstream starts to do that. It would give us more confident about using etcd3 as default backend.

**Comment by wojtek-t** [comment]
@xiang90 - I'm happy to setup tests, but to do that, we need to be able to configure it using kube-up startup scripts. And I also don't have time to do it anytime soon.

So it seems that we are blocked on changing kube-up to support etcd3.

**Comment by xiang90** [comment]
Do you know who maintains kube up?

**Comment by wojtek-t** [comment]
@xiang90 - there isn't a single person - anyone who needs to modify it is doing the work.

**Comment by xiang90** [comment]
@wojtek-t @timothysc OK. I am not aware of that before. I thought someone maintains the scripts. We will take some time to look into it next week. Thanks for the information.

**Comment by wojtek-t** [comment]
So let me put an update of it what our plan for it is.

- we are currently running with 2.2.1 version
- we are targetting 3.0.x with v3 protocol (or maybe even 3.1.x)
- the caveat is that we shouldn't skip any version between upgrades.
- so our plan to upgrade from 2.2.1:
    - backup 2.2.1 data
    - start 2.3.7 version for a moment
    - once it is healthy, kill it
    - start 3.0.x version for a moment
    - once it is healthy, kill it
    - do offline migration of data to v3 API
    - from this simply start 3.0.x version in v3 mode
- additionally, the etcd version and API mode will be stored in a dedicated file (so that if it fails because of some random reason after e.g. already starting 3.0.x in v2 mode, we won't start from the procedure beginning, but from that point)
- note that it means that the etcd image will have a bunch of different binaries built into it.
- using this strategy, we will rollout the target version in a single release (which is highly desirable for GKE)

The above will be implemented by modifying existing "migrate-if-needed.sh" script - it will have all those steps already built into it.

@lavalamp @fabioy @alex-mohr - if there is anything that you don't agree with (I think this captures what were talking about it), just let me know [obviously this doesn't describe testing procedure, which for GKE will be done internally).

**Comment by lavalamp** [comment]
I think we need to document an emergency dump-and-restore method to get back to an older etcd in the worst case.

Note that such a thing would be very unfortunate and require a full cluster reset, as ResourceVersions going backwards violates our locking guarantees.

We can extensively test the rollforward procedure, but it's harder to be sure we'll never find a bug in etcd that requires rolling back to a previous version.

**Comment by justinsb** [comment]
I agree with @lavalamp that we need a dump-and-restore procedure given the complexity of the upgrade procedure.

But then I'm wondering if we should just do a full logical data-dump (maybe even in our own code, if the etcd dump/restore utilities are tied to a particular version or we didn't want to assume they will always work), upgrade to the new version and then do a restore into etcd3.

I'd actually hope we could also use this code more generally - e.g. to perform online full backups of etcd, now that it supports MVCC.

If this is unfeasible at the etcd layer, there have also been requests for this to work at the kubectl layer (i.e. a `kubectl dump` which can be `kubectl apply`d).

**Comment by wojtek-t** [comment]
I definitely agree that we need to document dump-restore procedure, but this in my mind is a next step, and above I was just describing what to do with the container to enable "rollforward" testing.

@justinsb Regarding online backups - internally in GKE we are doing this (using standard etcd tools). Restore procedure is a bit more complicated, but we will try to work on it and create some script for open-source k8s.

**Comment by lavalamp** [comment]
Justin, it's not desirable to do the dump/restore pattern for upgrades if
we can possibly avoid it, since it loses ResourceVersion metadata. Since
components depend on RV being a logical clock that only ever increments,
we'd have to restart everything after doing anything that causes them to go
backwards.

kubectl has an export command.

dump/restore doesn't make a whole lot of sense at the kubectl level, IMO.
There's too much cluster-specific stuff in objects to make things portable
between clusters that way, and it's basically never an operation you'd want
to perform on a single cluster.

Honestly dump/restore doesn't make much sense even at the db level, since
even minutes or seconds of cluster operation could make dropping back to an
earlier state disastrous, and the larger the cluster the more likely some
tenant is doing something not very idempotent.

**Comment by timothysc** [comment]
closing now enabled by default.

**Comment by wojtek-t** [comment]
Reopening, since we don't have any documentation how to do an upgrade in OSS.

**Comment by timothysc** [comment]
@wojtek-t there is a docs ticket here: https://github.com/kubernetes/kubernetes.github.io/issues/2172

**Comment by wojtek-t** [comment]
OK - sorry. I must have missed it - too many emails when I was OOO.

**Comment by teotia39** [comment]
For Kubernetes 1.6.1 etcd 3 , why we are not able to access the "/registry" key, How to access this key ?

As this key is not present on default 2379 port , curl to http://127.0.0.1:2379/v1/keys/registry is failing , which was working very well in alpha version. How to access those kubernetes objects now.

**Comment by wojtek-t** [comment]
@teotia39 - with etcd v3, data are not accessible via json endpoint. You need to use etcd tooling for it (ETCDCTL_API=3 etcdctl).

**Comment by teotia39** [comment]
so that tool doesn't comes by default with kubeadm , and we have to explicitly install the etcdctl on top of kubeadm to access the endpoints ?

**Comment by wojtek-t** [comment]
this tool is built in the default etcd Docker image. You can just exec it from the container running etcd.

**Comment by teotia39** [comment]
and even if we install etcd cluster separately , and point that to k8's , still we wont be able to access the registered kubernetes objects via rest calls , and we still have to use etcdctl command line to get those objects.

**Comment by smarterclayton** [comment]
Etcd3 has a grpc endpoint, so no, no rest calls.
